observables.py

```diff
@@ -0,0 +1,90 @@
+"""
+observables.py
+
+functions for calculating observables on a stack of states
+"""
+from math import acosh, sqrt
+
+import matplotlib.pyplot as plt
+
+from geometry import get_shift
+
+def two_point_green_function(phi_states, x_0, x_1, length):
+    """Calculates the two point connected green function given a set of
+    states G(x) where x = (x_0, x_1)
+    """
+    shift = get_shift(
+        length,
+        shifts=((x_0, x_1),),
+        dims=((0, 1),),
+    ).view(-1) # make 1d
+
+    g_func = (
+        (phi_states[:, shift]*phi_states).mean(dim=0) - # mean over states
+        phi_states[:, shift].mean(dim=0)*phi_states.mean(dim=0)
+    )
+    return g_func.mean() # integrate over y and divide by volume
+
+def zero_momentum_green_function(phi_states, length):
+    """Calculate the zero momentum green function as a function of t which
+    is assumed to be in the first dimension
+    """
+    g_0_ts = []
+    for t in range(length):
+        g_0_t = 0
+        for x in range(length):
+            # not sure if factors are correct here should we account for
+            # forward-backward in green function?
+            g_0_t += float(two_point_green_function(phi_states, t, x, length))
+        g_0_ts.append(g_0_t)
+    return g_0_ts
+
+def effective_pole_mass(g_0_ts):
+    """Calculate the effective pole mass as a function of t, from t=1 to t=L-2"""
+    m_t = []
+    for i, g_0_t in enumerate(g_0_ts[1:-1], start=1):
+        m_t.append(acosh((g_0_ts[i-1] + g_0_ts[i+1])/(2*g_0_t)))
+    return m_t
+
+def susceptibility(g_0_ts):
+    """Calculate the susceptibility, which is the sum of two point connected
+    green functions over all seperations. Since we are calculating the zero
+    momentum green function we can just sum over t"""
+    return sum(g_0_ts)
```

Comment:
@tgiani do you agree with this? I don't know if I am saying something stupid but if we get zero momentum GF and sum on t, I think this should be correct?

Reply:
uhm do you want to implement Eq.(25) of the paper? I m not sure.. Who is ```length``` in ``` zero_momentum_green_function```? It should be the full size of the lattice, right?

Reply:
ah no right, sorry got it now

Reply:
I think is correct

Reply:
If I take eq. 23 and set p=0 and then take sum_t the question is do I get RHS of eq. 25

the length here is just used for get_shift not really used in the calculation

Reply:
To me it looks correct.. You have to sum over all the points ```x``` of the lattice and you can do this  by summing first over the spatial component and then over the time one

Reply:
ok good, then we should understand why I can get acceptance of like 45% but the numbers are way off

Reply:
Do you have numbers for L=2?

Reply:
no this is L=6, I can get L=2 pretty quickly later

Reply:
but what should I set M^2 and lambda to?

Reply:
just set them to 1, I ll use the same in the notebook

Reply:
ok with these I get:
```
Accepted: 100000, Rejected: 11351, Fraction: 0.9
Ising Energy: 0.057270217686891556
Susceptibility: 0.3175181970000267
```

Reply:
I m getting susceptibility ```0.348916```

Reply:
I don t have ising energy yet

Reply:
ok if I do more training and then add a random interval between samples when calculating the observables of 5 to account for some autocorrelation then I get
```
Accepted: 100000, Rejected: 6044, Fraction: 0.94
Ising Energy: 0.0637047290802002
Susceptibility: 0.3435881659388542
Time to run MC for a chain of 100000 samples on an L=2 lattice: 4.277069807052612 seconds
```
I think this looks like the training, sampling and susceptibility are at least ok

Reply:
In the most recent version of the [https://arxiv.org/pdf/1904.12072.pdf](paper) they corrected a mistake in equation 23, meaning we should be taking the mean, not sum, over `x` for each `g_0_t` in the zero momentum Green's function (and, as a consequence, the `susceptibility` function will have to be altered).

The agreement for L=2 susceptibility is really encouraging!

Reply:
the action is defined incorrectly in the arxiv version but the momentum green function is incorrect in the DOI.. So annoying 😆

Reply:
Just a double check: with my notebook I get the same number for susceptibility, and 0.0642849 for the Ising energy, so it looks good!